### DISCUSSION TITLE
Request: Allow deleting inactive query data in `queryCache.invalidateQueries`

### DISCUSSION THREAD

**Discussion by joealden** [TanStack/query#894] [General] [answered]
### Problem

As mentioned in #708, my teams wants to upgrade from v1 of `react-query`, but due to the change in behaviour from `queryCache.refetchQueries` -> `queryCache.invalidateQueries`, the path to do so hasn't been clear to us. The reason for this is because in a lot of our mutation `onSuccess` callbacks, we await the response of all of our `queryCache.refetchQueries` calls so that the mutation call itself only "finishes" when the data it effects has finished re-fetching.

This works in v1 because `queryCache.refetchQueries` would refetch inactive data, but `queryCache.invalidateQueries` doesn't unless you set the `refetchInactive` flag to `true`. While doing this would allow our team to migrate to v2 of `react-query`, it would still mean that we would be potentially overfetching a lot of data we don't need to be.

For this reason, we'd like a way to tell `react-query` to **delete** inactive queries data from the cache that match the specified query key, meaning that if/when this data is requested by a query in the future, it won't serve the content in the cache that we know for sure is stale, it will go out to the network and fetch it fresh and show loading the necessary loading states.

### Possible Solutions

A way we could see this being implemented in a non-breaking way would be to have a new `deleteInactive` boolean config option for  `queryCache.invalidateQueries` (that defaults to `false` like `refetchInactive` does). The problem with this would be that it'd be undefined behaviour (or not clear) if one was to have both `refetchInactive` and `deleteInactive` set to `true` at the same time.

A way we could see this being implemented in a likely better, but breaking way (v3?) would be to replace the `refetchInactive` config option with an option called something like `inactive` or `forInactive` that would have a string union type, not a boolean, where the options where something like `"invalidateOnly" | "refetch" | "delete"` where `"invalidateOnly"` is the default (I'm not sure `"invalidateOnly"` is the most descriptive name for this option though).

Let me know if you see any problems with what we've suggested, and if this is something that you'd be open to having added!

**Comment by tannerlinsley** [comment, marked as answer]
What about just removing the queries with a specified Query key that do not have any active instances/observers?

**Reply by joealden** [reply]
Good point, I didn't know we could find out in user-land what queries are currently not in use (via `queryCache.getQueries`).

Thanks!
